### PR TITLE
Visual testing fix: replace gemini-express by gemini-polyserve

### PR DESCRIPTION
**Pull request by limonte**
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-element-skeleton/24)

### CODE REVIEW

**Comment by manolo** [comment]
Doing this means that running `gulp perf:run` or `gemini update` wont work out-of-the-box, you have to previously launch `polyserve` in a separated session. :-1:

I would leave gulpfile and travis files as they was before, and remove the gemini-express dependency by gemini-polyserve which is a plugin I have just created to start/stop polyserve.

I have applied those changes to vaadin-button PR #2 and works perfectly.

---

Review status: 0 of 16 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24#-:-KdGdzD9fBLaZA567DOT:b-mzqqlc)*

**Comment by manolo** [comment]
Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/performance/config.json, line 2 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24#-KdV3-pIoo0xoH0n0jWU:-KdV3-pIoo0xoH0n0jWV:bado6uc) ([raw file](https://github.com/vaadin/vaadin-element-skeleton/blob/be9bcee1672ce3c2afe0e5fc26db01ca8253375d/test/performance/config.json#L2)):*
> ```JSON
> [
>   {"name": "index.html", "threshold": 0.75}
> ```

That is very unpredictable, yes. Not sure whether these test are really useful

---

*[test/visual/index.html, line 6 at r2](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24#-KdV2rnjmVXjf6RWSK64:-KdV2rnjmVXjf6RWSK65:b-emnez6) ([raw file](https://github.com/vaadin/vaadin-element-skeleton/blob/be9bcee1672ce3c2afe0e5fc26db01ca8253375d/test/visual/index.html#L6)):*
> ```HTML
>   <meta charset="UTF-8">
>   <title></title>
>   <script src="../../bower_components/webcomponentsjs/webcomponents-lite.js"></script>
> ```

It's needed for Edge and FF

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24)*

**Comment by manolo** [comment]
Review status: 0 of 14 files reviewed at latest revision, 1 unresolved discussion.

---

*[test/visual/test.js, line 1 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24#-KdV4pn4yIVa3QnJIWWh:-KdV4pn4yIVa3QnJIWWi:bsjlx99) ([raw file](https://github.com/vaadin/vaadin-element-skeleton/blob/be9bcee1672ce3c2afe0e5fc26db01ca8253375d/test/visual/test.js#L1)):*
> ```JavaScript
> gemini.suite('vaadin-element', function(rootSuite) {
> ```

Fo you mind to follow the same pattern than in vaadin input? That makes easy to have multiple visual tests in different files.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24)*

**Comment by limonte** [comment]
Review status: 0 of 14 files reviewed at latest revision, 1 unresolved discussion.

---

*[test/visual/test.js, line 1 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24#-KdV4pn4yIVa3QnJIWWh:-KdWCeNX8g-MLSPGI_7n:b-u9l45z) ([raw file](https://github.com/vaadin/vaadin-element-skeleton/blob/be9bcee1672ce3c2afe0e5fc26db01ca8253375d/test/visual/test.js#L1)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Fo you mind to follow the same pattern than in vaadin input? That makes easy to have multiple visual tests in different files.
</blockquote></details>

I'm not sure if we need to add any more complexity to the skeleton.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24)*

**Comment by manolo** [comment]
Reviewed 1 of 1 files at r1, 17 of 19 files at r2, 1 of 1 files at r3.
Review status: 13 of 14 files reviewed at latest revision, 1 unresolved discussion.

---

*[test/visual/test.js, line 1 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24#-KdV4pn4yIVa3QnJIWWh:-KdZxtW9rfZNE6Zt7nfN:b-rx286v) ([raw file](https://github.com/vaadin/vaadin-element-skeleton/blob/be9bcee1672ce3c2afe0e5fc26db01ca8253375d/test/visual/test.js#L1)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

I'm not sure if we need to add any more complexity to the skeleton.
</blockquote></details>


I don't see any complexity increment between these two blocks


```
gemini.suite('vaadin-element', function(rootSuite) {
  rootSuite.setUrl('/');

  gemini.suite('vaadin-element', function(suite) {
    suite
      .setCaptureElements('#visual-tests')
      .before(function(actions, find) {
        this.button = find('button');
      })
      .capture('normal-button', {}, function(actions, find) {
        actions.wait(3000); // hack for IE/Edge
      })
      .capture('clicked-button', function(actions) {
        actions.mouseDown(this.button);
      });
  });
 
});
```

```
gemini.suite('vaadin-input', function(rootSuite) {

  gemini.suite('default-tests', function(suite) {
    suite
      .setUrl('/default.html')
      .setCaptureElements('#default-tests')
      .before(function(actions, find) {
        this.button = find('button');
      })
      .capture('normal-button', {}, function(actions) {
        actions.wait(3000);
      })
      .capture('clicked-button', {}, function(actions) {
        actions.mouseDown(this.button);
        actions.wait(3000);
      }));
  });

});
```

So basically, what I ask for is
 - Don't set the Url in the root suite, so as we can have different files per certain test, because sometimes it's not possible to have tests in the same file, for instance when loading a theme that applies to all the page.
 - Use `default.html` instead of `index.html` so as in the test file we know exactly which file is being tested.
 - Note that in the second test you also have to wait to take the capture, it's added in the proposed block.
 - The wait hack is not only needed for Edge, some times the async loading of the webcomponents polyfill introduces a while in FF and Safari, so the comment could be removed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24)*

**Comment by limonte** [comment]
Done.

---

Review status: 6 of 16 files reviewed at latest revision, 1 unresolved discussion.

---

*[test/visual/test.js, line 1 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24#-KdV4pn4yIVa3QnJIWWh:-KddzMHF8qzNOIg7QXML:b-dtkds3) ([raw file](https://github.com/vaadin/vaadin-element-skeleton/blob/be9bcee1672ce3c2afe0e5fc26db01ca8253375d/test/visual/test.js#L1)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>


I don't see any complexity increment between these two blocks


```
gemini.suite('vaadin-element', function(rootSuite) {
  rootSuite.setUrl('/');

  gemini.suite('vaadin-element', function(suite) {
    suite
      .setCaptureElements('#visual-tests')
      .before(function(actions, find) {
        this.button = find('button');
      })
      .capture('normal-button', {}, function(actions, find) {
        actions.wait(3000); // hack for IE/Edge
      })
      .capture('clicked-button', function(actions) {
        actions.mouseDown(this.button);
      });
  });
 
});
```

```
gemini.suite('vaadin-input', function(rootSuite) {

  gemini.suite('default-tests', function(suite) {
    suite
      .setUrl('/default.html')
      .setCaptureElements('#default-tests')
      .before(function(actions, find) {
        this.button = find('button');
      })
      .capture('normal-button', {}, function(actions) {
        actions.wait(3000);
      })
      .capture('clicked-button', {}, function(actions) {
        actions.mouseDown(this.button);
        actions.wait(3000);
      }));
  });

});
```

So basically, what I ask for is
 - Don't set the Url in the root suite, so as we can have different files per certain test, because sometimes it's not possible to have tests in the same file, for instance when loading a theme that applies to all the page.
 - Use `default.html` instead of `index.html` so as in the test file we know exactly which file is being tested.
 - Note that in the second test you also have to wait to take the capture, it's added in the proposed block.
 - The wait hack is not only needed for Edge, some times the async loading of the webcomponents polyfill introduces a while in FF and Safari, so the comment could be removed
</blockquote></details>

Done, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24#-:-KddzEpumbOPE0F5VFEO:b-896fix)*

**Comment by manolo** [comment]
Review status: 6 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/visual/test.js, line 14 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24#-Kde-TfbvL0VwWd9cfoz:-Kde-TfcHbo8CTvSfEPV:bsllw2l) ([raw file](https://github.com/vaadin/vaadin-element-skeleton/blob/d3ecf0c490c837ead51f65b8d96225b597dde64a/test/visual/test.js#L14)):*
> actions.mouseDown(this.button);
Dont you need the actions.wait also here?, Not sure if before or after clicking button but I feel that we need this, probably before, so as when we click we already know that everything was rendered. 
Probably the first actions.wait in the previous test help but just to be sure

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24)*

**Comment by limonte** [comment]
Review status: 6 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/visual/test.js, line 14 at r4](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24#-Kde-TfbvL0VwWd9cfoz:-Kde0UtqtGMtFwHfHuke:bbr577x) ([raw file](https://github.com/vaadin/vaadin-element-skeleton/blob/d3ecf0c490c837ead51f65b8d96225b597dde64a/test/visual/test.js#L14)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

> actions.mouseDown(this.button);
Dont you need the actions.wait also here?, Not sure if before or after clicking button but I feel that we need this, probably before, so as when we click we already know that everything was rendered. 
Probably the first actions.wait in the previous test help but just to be sure
</blockquote></details>

No need to wait here. The visual result of mouse-down event is synchronous and available to capture right away.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24)*

**Comment by manolo** [comment]
:lgtm_strong: 

---

Reviewed 6 of 26 files at r2, 17 of 17 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-element-skeleton/24#-:-Kde3DSCAyQT-ZEtOp0v:b-uwkowd)*
